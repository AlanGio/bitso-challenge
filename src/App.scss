html,
body {
  height: 100%;
  min-width: 600px;
}

body {
  overflow: hidden;
  width: 100%;
  background-color: #eee;
  padding-top: 60px;
}

#root,
.App {
  height: 100%;
}

.App {
  margin: 0;
}

@media (max-width: 576px) {
  html,
  body {
    min-width: auto;
    height: 100%;
  }
  body {
    padding-top: 130px;
  }
}
