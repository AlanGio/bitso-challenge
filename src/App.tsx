import React, { ChangeEvent, useCallback, useState } from "react";

import { Controls } from "./components/Controls";
import { Grid } from "./components/Grid";

import { loadItems, detectIslands } from "./utils/utils";

import "./App.scss";

export const COL_INIT = 4;
export const ROW_INIT = COL_INIT;

const App = () => {
  const [columns, setColumns] = useState<number>(COL_INIT);
  const [rows, setRows] = useState<number>(ROW_INIT);
  const [cells, setCells] = useState<boolean[][]>(loadItems(columns, rows));

  const handleChangeColumns = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const columnsQty = +event.target.value;
      setColumns(columnsQty);
      setCells(loadItems(columnsQty, rows, cells));
    },
    [cells, rows]
  );

  const handleChangeRows = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const rowsQty = +event.target.value;
      setRows(rowsQty);
      setCells(loadItems(columns, rowsQty, cells));
    },
    [cells, columns]
  );

  const handleOnClickCell = useCallback(
    (row: number, column: number) => {
      let selectedCells = [...cells];
      selectedCells[row][column] = !selectedCells[row][column];
      setCells(selectedCells);
    },
    [cells]
  );

  const islands = detectIslands(columns, rows, cells);

  return (
    <div className="App">
      <Controls
        columns={columns}
        onChangeColumns={handleChangeColumns}
        onChangeRows={handleChangeRows}
        rows={rows}
        islands={islands}
      />
      <Grid cells={cells} onClickCell={handleOnClickCell} />
    </div>
  );
};

export default App;
