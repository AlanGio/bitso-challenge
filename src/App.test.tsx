import React from 'react';
import { render } from '@testing-library/react';
import App, { COL_INIT, ROW_INIT } from './App';

describe('<App />', () => {

  it('should render App component with default columns and rows', () => {
    const { container } = render(
      <App />
    );

    // Check ROWs quantity
    expect(container.querySelectorAll('.cell-row').length).toBe(ROW_INIT);
    // Check COLUMNS quantity
    expect(container.querySelectorAll('.cell-row:first-child > .cell').length).toBe(COL_INIT);
    // Check CELLs quantity
    expect(container.querySelectorAll('.cell').length).toBe(ROW_INIT * COL_INIT);

    expect(container).toMatchSnapshot();

  });

});
