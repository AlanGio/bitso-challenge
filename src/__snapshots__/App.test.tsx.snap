// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render App component with default columns and rows 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="component-controls"
    >
      <div
        class="controls-container container-fluid"
      >
        <form
          class=""
        >
          <div
            class="row"
          >
            <div
              class="col-sm-5 col-12"
            >
              <div
                class="form-group"
              >
                <label
                  class="form-label"
                  for="columnsRange"
                >
                  Columns 
                  4
                </label>
                <input
                  class="form-control-range"
                  id="columnsRange"
                  max="50"
                  min="2"
                  type="range"
                  value="4"
                />
              </div>
            </div>
            <div
              class="col-sm-5 col-10"
            >
              <div
                class="form-group"
              >
                <label
                  class="form-label"
                  for="rowsRange"
                >
                  Rows 
                  4
                </label>
                <input
                  class="form-control-range"
                  id="rowsRange"
                  max="50"
                  min="2"
                  type="range"
                  value="4"
                />
              </div>
            </div>
            <div
              class="col-sm-2 col-2"
            >
              Islands: 
              0
            </div>
          </div>
        </form>
      </div>
    </div>
    <div
      class="component-grid"
    >
      <div
        class="grid-container container-fluid"
      >
        <div
          class="cell-row row"
        >
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
        </div>
        <div
          class="cell-row row"
        >
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
        </div>
        <div
          class="cell-row row"
        >
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
        </div>
        <div
          class="cell-row row"
        >
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
          <div
            class="cell col"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
